File: output/gpt-5.4-2026-03-05/few/cpp/Burocracia.cpp
#include <bits/stdc++.h>
using namespace std;

struct Fenwick {
    int n;
    vector<int> bit;
    Fenwick() {}
    Fenwick(int n) { init(n); }
    void init(int n_) {
        n = n_;
        bit.assign(n + 1, 0);
    }
    void add(int idx, int val) {
        for (; idx <= n; idx += idx & -idx) bit[idx] += val;
    }
    int sumPrefix(int idx) const {
        int r = 0;
        for (; idx > 0; idx -= idx & -idx) r += bit[idx];
        return r;
    }
    int rangeSum(int l, int r) const {
        if (l > r) return 0;
        return sumPrefix(r) - sumPrefix(l - 1);
    }
};

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    int N;
    if (!(cin >> N)) return 0;

    vector<int> parent(N + 1, 0);
    vector<vector<int>> children(N + 1);
    parent[1] = 0;
    for (int i = 2; i <= N; i++) {
        cin >> parent[i];
        children[parent[i]].push_back(i);
    }

    vector<int> tin(N + 1), tout(N + 1), depth(N + 1);
    vector<vector<int>> byDepth(N + 1);

    int timer = 0;
    vector<pair<int,int>> st;
    st.reserve(2 * N);
    st.push_back({1, 0});
    depth[1] = 0;

    while (!st.empty()) {
        auto [u, state] = st.back();
        st.pop_back();
        if (state == 0) {
            tin[u] = ++timer;
            byDepth[depth[u]].push_back(tin[u]);
            st.push_back({u, 1});
            for (int i = (int)children[u].size() - 1; i >= 0; i--) {
                int v = children[u][i];
                depth[v] = depth[u] + 1;
                st.push_back({v, 0});
            }
        } else {
            tout[u] = timer;
        }
    }

    int LOG = 1;
    while ((1 << LOG) <= N) LOG++;
    vector<vector<int>> up(LOG, vector<int>(N + 1, 0));
    for (int i = 1; i <= N; i++) up[0][i] = parent[i];
    for (int j = 1; j < LOG; j++) {
        for (int i = 1; i <= N; i++) {
            up[j][i] = up[j - 1][ up[j - 1][i] ];
        }
    }

    auto kthAncestorOriginal = [&](int v, int k) {
        for (int j = 0; j < LOG; j++) {
            if (k & (1 << j)) v = up[j][v];
        }
        return v;
    };

    int Q;
    cin >> Q;

    Fenwick bit(N);
    vector<int> restructured(N + 1, 0);

    auto isRestructuredOnPathStrict = [&](int a, int b) {
        if (depth[a] + 1 > depth[b]) return false;
        int d = depth[b] - depth[a];
        auto &vec = byDepth[depth[a] + 1];
        int l = lower_bound(vec.begin(), vec.end(), tin[a]) - vec.begin();
        int r = upper_bound(vec.begin(), vec.end(), tout[a]) - vec.begin();
        if (l >= r) return false;
        int cnt = bit.rangeSum(l + 1, r);
        return cnt > 0;
    };

    string out;
    out.reserve(Q * 3);

    for (int qi = 0; qi < Q; qi++) {
        int type;
        cin >> type;
        if (type == 1) {
            int v, k;
            cin >> v >> k;

            int targetDepth = depth[v] - k;
            int lo = 0, hi = k, ans = 0;
            while (lo <= hi) {
                int mid = (lo + hi) >> 1;
                int anc = kthAncestorOriginal(v, mid);
                if (!isRestructuredOnPathStrict(anc, v)) {
                    ans = mid;
                    lo = mid + 1;
                } else {
                    hi = mid - 1;
                }
            }

            int x = kthAncestorOriginal(v, ans);
            if (ans == k) {
                out += to_string(x);
                out += '\n';
            } else {
                int rem = k - ans;
                int y = kthAncestorOriginal(v, ans + 1);
                int cur = y;
                if (rem == 1) {
                    out += to_string(cur);
                    out += '\n';
                } else {
                    int finalAns = kthAncestorOriginal(cur, rem - 1);
                    out += to_string(finalAns);
                    out += '\n';
                }
            }
        } else {
            int v;
            cin >> v;
            if (v != 1 && !restructured[v]) {
                restructured[v] = 1;
                int d = depth[v];
                int pos = lower_bound(byDepth[d].begin(), byDepth[d].end(), tin[v]) - byDepth[d].begin() + 1;
                bit.add(pos, 1);
            }
        }
    }

    cout << out;
    return 0;
}

File: output/gpt-5.4-2026-03-05/few/python/Burocracia.py
import sys

sys.setrecursionlimit(200000)

data = list(map(int, sys.stdin.read().split()))
if not data:
    exit()

ptr = 0
n = data[ptr]
ptr += 1

parent0 = [0] * (n + 1)
children = [[] for _ in range(n + 1)]
for i in range(2, n + 1):
    p = data[ptr]
    ptr += 1
    parent0[i] = p
    children[p].append(i)

tin = [0] * (n + 1)
tout = [0] * (n + 1)
depth0 = [0] * (n + 1)
order = []

stack = [(1, 0)]
timer = 0
while stack:
    v, state = stack.pop()
    if state == 0:
        tin[v] = timer
        order.append(v)
        timer += 1
        stack.append((v, 1))
        ch = children[v]
        for i in range(len(ch) - 1, -1, -1):
            u = ch[i]
            depth0[u] = depth0[v] + 1
            stack.append((u, 0))
    else:
        tout[v] = timer - 1

q = data[ptr]
ptr += 1

ops = []
restruct = []
for _ in range(q):
    t = data[ptr]
    ptr += 1
    if t == 1:
        v = data[ptr]
        k = data[ptr + 1]
        ptr += 2
        ops.append((1, v, k))
    else:
        v = data[ptr]
        ptr += 1
        ops.append((2, v))
        restruct.append(v)

m = len(restruct)
if m == 0:
    LOG = (n + 1).bit_length()
    up = [[0] * (n + 1) for _ in range(LOG)]
    up[0] = parent0[:]
    for j in range(1, LOG):
        prev = up[j - 1]
        cur = up[j]
        for i in range(1, n + 1):
            cur[i] = prev[prev[i]]

    out = []
    for op in ops:
        if op[0] == 1:
            v, k = op[1], op[2]
            b = 0
            while k:
                if k & 1:
                    v = up[b][v]
                k >>= 1
                b += 1
            out.append(str(v))
    sys.stdout.write("\n".join(out))
    exit()

restruct_sorted = sorted(set(restruct), key=lambda x: tin[x])
roots = []
for v in restruct_sorted:
    if not roots or tout[roots[-1]] < tin[v]:
        roots.append(v)

root_id = [-1] * (n + 1)
root_depth = [0] * (n + 1)
for rid, r in enumerate(roots):
    d = depth0[r]
    l = tin[r]
    rr = tout[r]
    for pos in range(l, rr + 1):
        u = order[pos]
        root_id[u] = rid
        root_depth[u] = d

bit = [0] * (m + 2)

def bit_add(i, delta):
    i += 1
    while i <= m + 1:
        bit[i] += delta
        i += i & -i

def bit_sum(i):
    s = 0
    i += 1
    while i > 0:
        s += bit[i]
        i -= i & -i
    return s

events = [[] for _ in range(q + 1)]
for idx_r, v in enumerate(restruct):
    rid = root_id[v]
    if rid != -1 and roots[rid] == v:
        events[idx_r + 1].append(rid)

queries = []
for idx_op, op in enumerate(ops, 1):
    if op[0] == 1:
        v, k = op[1], op[2]
        rid = root_id[v]
        if rid == -1:
            queries.append((idx_op, v, k, -1))
        else:
            queries.append((idx_op, v, k, rid))

ans_depth = [0] * len(queries)

qi = 0
for t in range(1, q + 1):
    for rid in events[t]:
        bit_add(rid, 1)
    while qi < len(queries) and queries[qi][0] == t:
        _, v, k, rid = queries[qi]
        if rid == -1:
            ans_depth[qi] = depth0[v] - k
        else:
            cnt = bit_sum(rid)
            droot = root_depth[v]
            if k <= depth0[v] - droot:
                ans_depth[qi] = depth0[v] - k
            else:
                rem = k - (depth0[v] - droot)
                ans_depth[qi] = droot - (rem - 1) - cnt
        qi += 1

by_depth = [[] for _ in range(n + 1)]
for v in range(1, n + 1):
    by_depth[depth0[v]].append(v)

maxd = max(depth0)
for d in range(maxd + 1):
    by_depth[d].sort(key=lambda x: tin[x])

def find_descendant_at_depth(v, target_depth):
    arr = by_depth[target_depth]
    l = 0
    r = len(arr)
    tv = tin[v]
    while l < r:
        mid = (l + r) >> 1
        if tin[arr[mid]] < tv:
            l = mid + 1
        else:
            r = mid
    return arr[l]

out = []
qi = 0
for op in ops:
    if op[0] == 1:
        v = op[1]
        d = ans_depth[qi]
        out.append(str(find_descendant_at_depth(v, d)))
        qi += 1

sys.stdout.write("\n".join(out))